[FLINK-16536][network][checkpointing] Implement InputChannel state recovery for unaligned checkpoint

zhijiangW opened the pull request:
## What is the purpose of the change

During recovery process for unaligned checkpoint, the remote input channel state should also be recovered besides with existing operator states.
    
The remote channel would request buffer and then interact with ChannelStateReader to fill in the state data.  The filled buffer would be inserted into respective data queue for processing in normal way.

It should guarantee that the new data from upstream partition should not overtake the input state data to avoid mis-order issue.

## Brief change log

  - *Delay request partition in `RemoteInputChannel` after finishing reading channel state.*
  - *Add related interfaces method from input channel to handler stacks to notify the read of channel state*
  - *Implement the main process to read channel state via interacting with `ChannelStateReader`*

## Verifying this change

Add some new unit tests to verify the changes

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhijiangW commented:
Some thoughts to share during implementation. 

1. Which kind of thread should read the channel state? 
We can not use the task thread, because it will process and consume the filled buffers to recycle them to load more channel states. In other words, the task thread and state loading thread should work in parallel. 

If we introduce another thread pool  for reading state, it has more troubles. First we need to maintain the lifecycle of it. Second it will bring more threads overhead in a TaskExecutor container. Third it is hard to determine how many threads are suitable based on demands. One thread might not be enough for serving many input channels at the same time. Fourth it bring the thread model more complex besides with both task and netty threads.

Therefore, the existing netty threads are the most ideally candidate to solve all the above concerns. We do not need to worry about the netty thread is involving in IO operations, because it already exists the similar case on upstream side to read blocking partitions by netty thread. And we can also make use of the default load balance among input channels with netty threads.

2. The current implementation will delay partition request until finishing channel state recovery. It is mostly easy to avoid data order issues. But in future if we want to perform new checkpoints during recovery, the partition request should also be done meanwhile, and it would bring some difficulties.

One option is we give 0 credits during partition request to avoid receiving any data during recovery, then we need to adjust the logic when to announce the real credit again after recovery. But another trouble is that we can also receive `EndOfPartitionEvent` besides `CheckpointBarrier` without credits. So how to store this event in buffer queue to avoid mis-order issue.

Another potential issue is how to switch between reading channel state and receiving network data in netty thread if we want to support checkpoint during recovery. So I take the simple way in MVP now, to further think through how to support checkpoint during recovery future.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -311,10 +367,16 @@ public void recycle(MemorySegment segment) {
 
-		if (numAddedBuffers > 0 && unannouncedCredit.getAndAdd(numAddedBuffers) == 0) {
+		if (shouldNotifyChannelState) {
+			notifyReadChannelState();
```
`notifyStateBuffersAvailable()`? `notifyBuffersForReadChannelStateAvailable`?

zhijiangW replied:
I like `notifyStateBuffersAvailable`

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -149,16 +154,58 @@ void assignExclusiveSegments() throws IOException {
+				}
+			}
+
+			ChannelStateReader.ReadResult result = inputGate.stateReader.readInputData(channelInfo, buffer);
```
`stateReader.readInputData` is not thread safe as it is currently and this `readInputChannelState` method can be executed by netty threads concurrently for different channels, right?

zhijiangW replied:
Yes, `readInputChannelState` can be executed concurrently by multiple netty threads for different channels by design. I think in general task processing should be more faster than reading states, so one thread might not be enough for filling buffer to feed task thread well. And every channel actually has exclusive buffers which can be used in parallel to speed up recovery process.

I overlooked the `NotThreadSafe` annotation in `ChanelStateReaderImpl`.  Since every input channel handle will actually generate a separate `ChannelStateStreamReader` and respective stream, I was supposed one input channel state should not be read by multiple threads, but different channel states can be read by different threads concurrent. I would further confirm with Roman whether there are other limitations.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -95,7 +95,7 @@ public LocalInputChannel(
 
 	@Override
-	protected void requestSubpartition(int subpartitionIndex) throws IOException, InterruptedException {
+	protected void requestSubpartition() throws IOException, InterruptedException {
```
I actually considered whether to extract it as a separate commit or not when implementation. I think both ways can be interpreted from different motivations. 

If making it as a separate hotfix or refactor commit, we only save one argument from this interface method and fetch it from `SingleInputGate` instead. Then the benefit seems not very obvious if without the following input recovery change.

If making it along with the input recovery changes, the motivation seems a must-work to do. Since we migrate the actual usage of this argument outside of this method, then this argument should also be cleaned up as a result.

Therefore I preferred a bit not to make it a separate commit before. It is fine for my either ways actually. If you are in favor of separate commit, i am willing to change it.

pnowojski replied:
It made the final commit a bit smaller, so I think it was worth it (to speed up reviewing). As hotfix on it's own it's also a nice simplification I think.

zhijiangW commented:
Thanks for the review and suggestions @pnowojski !

Regarding the issue of which thread handling read state, I also fully considered your concerns of netty thread before. Let me further elaborate a bit:

> Netty pool size is configured for handling network traffic, not for reading data from the spilled files. They > should be configured separately, as 32 threads for netty is not that uncommon, while that many threads > for File IO doesn't make much sense.

The default number of netty thread is bounded to slot amount if not config explicitly. I also mentioned about the consideration of how many threads are proper for serving state read in previous comment https://github.com/apache/flink/pull/11687#issuecomment-611465439. Firstly I think one thread should probably be not enough is practice, the reason explained here https://github.com/apache/flink/pull/11687#discussion_r407222896. IMO 10 threads might be an ideally amount in SATA disks, and for SDD we can have more threads concurrent reading in general. Actually the proper size is also related to the current IO pressure, so it is hard to determine beforehand, then I am willing to leave the choice to exiting netty threads to avoid bringing additional thread cost.

It might seem a bit tricky to make netty thread not only work for network traffic by design. But it already exists the same scenarios now that netty thread would read spilled file for blocking partition in batch jobs, even the case is almost the same with reading channel state to some extent. From another aspect, the netty threads on consumer side take the role of feeding data for task processing either from network or from state files. If we do not use netty threads, they are almost idle during state recovery, except receiving new checkpoint barrier occasionally for future supporting. 

> Reading File IO is blocking operation. Blocking Netty threads can cause problems. 

In general I am always against blocking operation in netty thread. But for this case the netty threads should be almost idle during recovery as I mentioned above. And I also considered to support receiving new checkpoint barrier during recovery future, which makes netty thread also work for network read/write ops by internal (e.g. after reading every state buffer). If it takes several minutes to read one buffer in extremely environment to delay new checkpoint receive, I think it should be a minor issue. Because in such bad IO situation, the total recovery might take hours to finish, then supporting minutes level progress for new unaligned checkpoint in this period should be enough. In other words, it seems meaningless if we still support second level progress for new checkpoint during recovery, because the state consumption probably has not been moved forward yet.

In general I like your ideas of the possibility to implement the input recovery from another layer, which I have not thought of before. I agree that it might be better to not touch the existing `RemoteInputChannel` if possible because it is already involved in many things. I would further think though how many efforts we need to pay if make it work on separate `StreamTaskSpilledInput `/`SpilledInputGate`/`SpilledInputChannel` level or whether it is feasible to work in cleaner way, and feedback to you later.

zhijiangW commented:
Considering another possible option to implement the input recovery, I thought through a bit for the proposal of introducing a parallel `StreamTaskSpilledInput` /`SpilledInputGate`/`SpilledInputChannel `. If we want to further support the new unaligned checkpoint progress during recovery, that means two separate components should run in parallel future. Then it seems harder to control and bring more component changes if we work on more high-level components.

Regarding the lowest-level component `SpilledInputChannel`, I also think it should be more complex if running in parallel with respective `RemoteInputChannel` future for below concerns.

- `SingleInputGate` would maintain two suits of input channel collections. The normal remote/local channel collection can only receive events like `CheckpointBarrier`, `EndOfPartition`  during recovery, and the other `SpilledInputChannel` collection can read state. We might need to adjust the queue `inputChannelsWithData` in gate to control the priority of data from different channel collection.

- How to handle the exclusive/floating buffer distribution inside `SpilledInputChannel`? There might be two options. If we want to recovery only one channel state at a time by one thread, then we can give all the available buffers(regardless of exclusive/floating) to this channel. And after finishing this channel, we can move forward to the next channel with all available buffers. If we want o recovery multiple channels in parallel with different threads(as this PR did now), we might probably back to the existing mechanisms as in `RemoteInputChannel` , then we need to consider the code duplication issue.

- We still need to adjust the internal logic inside `RemoteInputChannel` to do so. E.g. when it receives the barrier from network, it should not request floating buffers based on backlog, along with the related logics with exclusive buffer and floating buffers, until all the `SpilledInputChannel` finish. 

- When all the `SpilledInputChannel` finish, it should notify the `RemoteInputChannel` restore the normal exclusive/floating buffers logic. The interaction is needed among different channels via `SingleInputGate`.

In conclusion, the essential difference with current PR implementation is whether to extract the logic `readChannelState` out of `RemoteInputChannel` to relief its load. In contrast, this logic is introduced into new `SpilledInputChannel` which would bring extra complexity in `SingleInputGate` component and the trouble for reusing the existing buffer distribution/request logic.

So I a bit prefer the current way in PR at-least for MVP now until we fully consider well for everything future.
But we can make some partial changes to control the channel state process on `SingleInputGate` component. E.g. introduce a separate thread pool to execute the channel read instead of reusing netty threads as above discussed, and control the channel read one by one instead of concurrently if have the requirement.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -243,7 +249,23 @@ void requestPartitions() throws IOException, InterruptedException {
+				for (InputChannel inputChannel : inputChannels.values()) {
+					executor.submit(() -> {
+						try {
+							inputChannel.requestSubpartition();
```
Could it be that we start reading from a channel before we actually request subpartition? For example:
```
Optional<BufferAndAvailability> RemoteInputChannel#getNextBuffer() throws IOException {
	(...)
	checkState(partitionRequestClient != null, "Queried for a buffer before requesting a queue."); // ??
```

or what if `partitionRequestClient` is already not null, but 
```
partitionRequestClient.requestSubpartition(partitionId, subpartitionIndex, this, 0);
```
hasn't completed?

zhijiangW replied:
Yes, it has another issue here.  In order to compatible with previous process, we can divide the previous `requestSubpartition` into two steps. The first step is to create client which can be done during `InputGate#setup` as before. The second step is to actual request partition which is triggered after reading channel state as done in this PR change. 

To do so, this check is still valid, but I guess it is still breaking the previous assumption which actually wants to guarantee the partition is really requested before calling `getNextBuffer`. But now the existing client does not mean the partition was requested already. So another option might remove this `checkState`.

zhijiangW replied:
I currently remove this `checkState` in PR for simple. If you have other thoughts, then we can further sync afterwards.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -243,7 +249,23 @@ void requestPartitions() throws IOException, InterruptedException {
+						} catch (Throwable t) {
+							inputChannel.setError(t);
+						}
+					});
```
I would move the for loop inside the executor, to avoid race conditions and not depend on `executor` being single threaded.

zhijiangW replied:
good point

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -149,6 +149,47 @@ void assignExclusiveSegments() throws IOException {
+			}
+
+			if (buffer == null) {
+				wait(10);
```
Is someone notifing this thread to wake up? Can not we wait on the buffer in another way?

zhijiangW replied:
TBH the current wait way is also not in my favor as I mentioned as unsatisfied points, because how long time to wait might be an issue.

I also considered other ways, but the key problem is that we can not exit this runnable to let the thread execute other runables in advance, so it seems somehow as blocking way here. In contrast, the previous floating request from netty thread was non-blocking way, so we do not need the mechanism of `notifyAll`  inside `RemoteInputChannel#recycle` and `RemoteInputChannel#notifyBufferAvailable`.

Maybe it is possible to add `notifyAll` in above two methods for compatible with the new process. WDYT?

pnowojski replied:
Ok I see. We can not block the shared executor for unspecified long time. This method would have to be rewritten in a such way, that once buffers are not available, it exits, and re-enqueuing action is done once more buffers are ready.

If we spawned our custom thread for unspilling, we could block it however we want, but in that case we would also need a mechanism to wake it up once more buffers are ready, instead of sleeping for fixed amount of time.

zhijiangW replied:
Yes, we have the same issue in another option of `SpilledInputChannel`/`SpilledInputGate`.

In conclusion, there may be several solutions:

- `wait()` if unavailable buffers: block the unspilling thread always and it fits our current requirements. We should not exit current channel to switch another channel temporarily, which might bring random IO. And it requires the wakeup mechanism when buffer available again.

- `wait(timeout)` if unavailable buffers: more or less the same with above `wait()`, but wakeup mechanism is not a mandatory, can be regarded as  somehow improvement to wakeup eerily. 

- unblocking way: terminate the current channel and unspill another channel with available buffers. It would bring random IO as mentioned above and not the current suggestion.

Based on the current situation with custom thread for unspilling, and allow only one thread to unspill channel one by one to avoid random IO, so I choose the unblocking option 1 or 2. 

Regarding the option 2, wakeup is not necessary, then we do not need to touch the previous processes `RemoteInputChannel#recycle` and `RemoteInputChannel#notifyBufferAvailable`. If we want to add the wakeup mechanism, option 1 also makes sense.

pnowojski replied:
Oh, I missed that we are already using a separate thread pool. In that case we don't need the not blocking option now.

It would be still good to get rid of timed wait in favour of some notifications though.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -213,19 +215,23 @@ public SingleInputGate(
+	}
+
 
-		requestPartitions();
```
who is now calling this method in batch tasks?

zhijiangW replied:
No matter streaming or batch, this method is always called by task thread when the `SingleInputGate` is created to setup. The only difference between streaming and batch is for RPC call `updateInputChannel` which is only valid in batch. But for batch we do not support unaligned cp, so the new transformed remote channel from unknown can request partition directly.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -289,6 +292,8 @@ protected StreamTask(
 			this.timerService = timerService;
 		}
+
+		this.ioExecutor = Executors.newSingleThreadExecutor(new ExecutorThreadFactory("state-io"));
```
rename `channelStateUnspillingExecutor`?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -213,19 +215,23 @@ public SingleInputGate(
 
 	@Override
-	public void setup() throws IOException, InterruptedException {
+	public void setup() throws IOException {
```
could we pass the state recovery parameters here? And have a single step `setup` method instead of introducing 2nd initialisation step (`setup` and `initializeState`).

zhijiangW replied:
I agree that it has benefits to reduce steps for managing the lifecycle of input gate. But I thought of another potential concern if integrating it within `setup`. 

The previous assumption was that the output recovery should execute earlier than the input recovery in order to occupy more floating buffers from global, to get the benefit to speedup recovery process. Now the output recovery is executed by task thread during `StreamTask#beforeInvoke`. If we integrate the input recovery within `setup` process, then it would perform before output recovery to occupy more floating buffers in limited buffers environment.

I also considered integrating the output recovery within `ResultPartition#setup`, but the output recovery is executed by task thread then it would block the following operations, so I wonder it might also bring potential risks to do so.

pnowojski replied:
> The previous assumption was that the output recovery should execute earlier than the input recovery in order to occupy more floating buffers from global, to get the benefit to speedup recovery process. Now the output recovery is executed by task thread during StreamTask#beforeInvoke. If we integrate the input recovery within setup process, then it would perform before output recovery to occupy more floating buffers in limited buffers environment.

Ok, I haven't considered that. I guess it's best to keep it as it is?

zhijiangW replied:
Yes, I guess so.

I found another benefit to execute the input state recovery inside `StreamTask` instead of `Task` stack(setup integration). The new dedicated executor for unspilling channel state is only meaningful for the unaligned checkpoint mode. So I tried not to create this executor for normal checkpoints, also for batch jobs to reduce total threads in TaskManager. This improvement can only be done inside `StreamTask` because it can judge the checkpoint mode via `StreamConfig`. And task class can not get this option because of `runtime` module can not depend on `streaming` module.

zhijiangW commented:
@flinkbot run travis

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
@flinkbot run travis

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -162,6 +163,95 @@ void assignExclusiveSegments() throws IOException {
+		}
+	}
+
+	private void beforeReadRecoveredState() {
```
`initializeCreditsForRecoveringState`?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -222,19 +225,65 @@ public SingleInputGate(
 
-		requestPartitions();
+	@Override
+	public void initializeStateAndRequestPartitions(
+			boolean hasStates,
+			@Nullable ExecutorService executor,
+			ChannelStateReader reader) throws Exception {
+
+		if (hasStates) {
+			checkNotNull(executor);
+			readRecoveredStateBeforeRequestPartition(executor, reader);
+		} else {
+			requestPartitions();
+		}
```
This doesn't look right - boolean `hasStates` and `@Nullable executor` - it looks like this check should be on a different layer. As this method looks like it's called only in one place, shouldn't it be just inlined?

zhijiangW replied:
It actually inlines two requirements on purpose. TBH I was also a bit torn to do so before.

My previous thought was trying not to affect the normal process if not enabling unaligned mode, that means requesting partitions directly, otherwise short lifecycle jobs might be sensitive for the potential performance regression with unnecessary state recovery process. And passing `nullable` executor can also simplify the unit tests without unaligned mode not to maintain the useless `executor`.

Another solution is to separate it into two methods. One is for `#requestPartitions` and the other is for `readRecoveredState`. But that means we also need to define another explicit interface method `InputGate#requestPartitions`, which would bring more tiny steps in the lifecycle of `InputGate`, like `setup` -> `readRecoveredState` -> `requestPartitions`. So I chose the inlined way to reduce this overhead.

pnowojski replied:
Ok, I see. This is messy :/ It comes down to this issue that input gates are initialised in the `Task` and state is handled in `StreamTask` :/ It should have been handled together upon the construction of `SingleInputGate` class.

I think I would be +0.1 for dropping `initializeStateAndRequestPartitions` and both `readRecoveredStateBeforeRequestPartition` and `requestPartitions` to the interfaces, but if you think it's better to have a single method, I'm fine with that.

zhijiangW replied:
I am planning to separate this inlined method for your 0.1 favor. :)

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -204,6 +205,9 @@
 
 	protected final MailboxProcessor mailboxProcessor;
 
+	@Nullable
+	private final ExecutorService channelStateUnspillingExecutor;
```
Add a `TODO` that it should be replaced by a global TaskManager ioExecutor?

zhijiangW replied:
Maybe, not quite sure whether the existing `ioExecutor` in `TaskManagerService` can be used for unspill  state purpose. Anyway, we can always introduce a dedicated executor on `TaskManager` level future.

zhijiangW commented on flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java:
```diff
@@ -914,7 +914,9 @@ public void testInitializeResultPartitionState() throws Exception {
 		mockEnvironment.addOutputs(Arrays.asList(partitions));
-		StreamTask task = new MockStreamTaskBuilder(mockEnvironment).build();
+		StreamConfig config = new StreamConfig(new Configuration());
+		config.setUnalignedCheckpointsEnabled(true);
```
The current default value for `ExecutionCheckpointingOptions#ENABLE_UNALIGNED` is still false. But in `StreamTask#beforeInvoke` the channel state recovery is only called when enabling unaligned mode. So ATM we still need this explicit setting for making the test through.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java:
```diff
@@ -131,7 +135,12 @@
-	public abstract void setup() throws IOException, InterruptedException;
+	public abstract void setup() throws IOException;
+
+	public abstract void initializeStateAndRequestPartitions(
```
`initializeState` is for alignment with the other side `ResultPartition#initializeState` introduced before. I am both fine with either naming.

pnowojski replied:
ok, let's keep it consistent. `initializeState` sounds to me more  like we are trying to initialize a state and keep using it later on (like in operators). `readState` better carries out the intention that the state is only used once.

zhijiangW replied:
Agree, I already renamed it on input side, and I can also refactor the partition side in a separate hotfix commit.

zhijiangW commented:
Thanks for further reviews @pnowojski ! 

I addressed some partial comments and updated the codes, only left the unresolved conversations to be further confirmed. After confirming all, i will further supplement some unit tests for the changes related to `SingleInputGate`.

zhijiangW commented:
@pnowojski I have added all the necessary unit tests for covering the changes inside `StreamTask`, `SingleInputGate` and `RemoteInputChannel`. I guess this [FLINK-17218](https://issues.apache.org/jira/browse/FLINK-17218) will cover the e2e ITCase for verifying the whole process of unaligned checkpoint with data correctness, then i will not introduce it in this PR.

pnowojski commented:
I think in the general concept code looks good to me, taking into account the trade offs we were discussing.

There are still some test failures, like dead locks. When you will be fixing test failures, could you keep them as separate fixups for me to review?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGate.java:
```diff
@@ -222,19 +226,54 @@ public SingleInputGate(
+		internalRequestPartitions(() -> executor.submit(() -> {
+
+			for (InputChannel inputChannel : inputChannels.values()) {
+				if (inputChannel instanceof RemoteInputChannel) {
+					try {
+						((RemoteInputChannel) inputChannel).readRecoveredState(reader);
+					} catch (Throwable t) {
+						inputChannel.setError(t);
+						return;
+					}
+				}
```
I think this block of code is forgetting about `getUnconsumedBuffer()` from the `LocalInputChannel` that was spilled as part of the input data.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/reader/AbstractRecordReader.java:
```diff
@@ -63,6 +63,8 @@ protected AbstractRecordReader(InputGate inputGate, String[] tmpDirectories) {
 	}
 
 	protected boolean getNextRecord(T target) throws IOException, InterruptedException {
+		inputGate.requestPartitions();
```
Why do we have to `requestPartitions()` once per every record?

zhijiangW replied:
This is the only concentrated place work for all the batch cases. Otherwise we have to add this call for all the specific invokable instances derived from `BatchTask` class, even we also need to consider many other cases which bypassed the invokable class and use the wrapped `AbstractRecordReader` directly in unit tests, etc.

I remembered in the early version, the `requestPartitions` was also placed inside `SingleInputGate#getNext`  method. Because of the mailbox requirement, it was migrated into `#setup` afterwards.

Considering the batch case unrelated to mailbox path, so i think it might be accepted to redo it still in the original place only work for batch cases.

pnowojski replied:
Can not we keep the previous setup logic if unaligned checkpoints are disabled/not configured (that would include batch?)? And add a `checkState` somewhere, that unaligned checkpoints can not be used without `StreamTask`/in streaming or something like that?

zhijiangW replied:
agree, it is not a perfect way, but the feasible simple way ATM to not maintain many different code paths, also not sensitive for batch code path.

pnowojski replied:
Could you explain why do you think:
> also not sensitive for batch code path.

?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/InputGate.java:
```diff
@@ -131,7 +133,18 @@
 	 * Setup gate, potentially heavy-weight, blocking operation comparing to just creation.
 	 */
-	public abstract void setup() throws IOException, InterruptedException;
+	public abstract void setup() throws IOException;
+
+	/**
+	 * It is only performed for unaligned checkpoint mode together with internal requesting partitions afterwards.
+	 * Otherwise only {@link #requestPartitions()} is performed for other checkpoint modes.
+	 *
+	 * @param executor the dedicated executor for performing the recovery state for all the internal channels.
+	 * @param reader the dedicated reader for unspilling the respective channel state from stored snapshots.
+	 */
+	public abstract void readRecoveredState(ExecutorService executor, ChannelStateReader reader) throws IOException;
+
+	public abstract void requestPartitions() throws IOException;
```
`requestPartitions()` requires similar java doc as `readRecoveredState()` explaining when it should be called.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -105,12 +98,15 @@ public RemoteInputChannel(
 		this.connectionId = checkNotNull(connectionId);
 		this.connectionManager = checkNotNull(connectionManager);
-		this.bufferManager = new BufferManager(this, 0);
+		// In theory it should get the total number of states to indicate the numRequiredBuffers.
+		// Since we can not get this information in advance, and considering only one input channel
+		// will read state at the same time by design, then we give a maximum value here to reduce
+		// unnecessary interactions with buffer pool during recovery.
+		this.bufferManager = new BufferManager(this, Integer.MAX_VALUE);
```
`Integer.MAX_VALUE`? Doesn't it mean that all floating buffers will be stuck permanently in on `RemoteInputChannel` (that happened to ask for them first?)

zhijiangW replied:
As I explained for `LocalInputChannel` case, this `numRequiredBuffers` setting is only for a bit optimization, actually we can unify them as 0 and adjust it while really requesting floating buffers in process. 

ATM we only have one input channel under unspill, so it makes sense to grab all the available floating buffers for this channel now. After this channel finishes unspilling,  then it would release all the floating buffers back to `LocalBufferPool` to be reused by other unspill channel.

There was a bit tricky to design the factor of `numRequiredBuffers` before. If one exclusive buffer is recycled or a floating buffer is recycled to notify available for the listener, it would double check whether the current listener still needs more floating buffers ATM based on `numRequiredBuffers`. If not needed, then the floating buffer would be return back to local pool to assign other listeners. 

For input channel unspill case, we can assume that the current channel is always needing more floating buffers until finish, to avoid the floating buffer back to local pool and request from pool again when need it next time.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -321,111 +274,43 @@ private void notifyCreditAvailable() {
 	}
 
 	public int getNumberOfRequiredBuffers() {
-		return numRequiredBuffers;
+		return bufferManager.getNumberOfRequiredBuffers();
 	}
```
Yes, it could be as you said. Actually it should be a separate hotfix if we want to fix it in this PR.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -0,0 +1,356 @@
+	// Getter properties
+	// ------------------------------------------------------------------------
+
+	int getNumberOfRequiredBuffers() {
```
`unsychronized` makes sense, but not for `@VisibleForTesting` because it is also used in formal codes of `RemoteInputChannel`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -321,111 +274,43 @@ private void notifyCreditAvailable() {
+		return bufferManager.getNumberOfRequiredBuffers();
 	}
 
 	public int getSenderBacklog() {
```
Exactly as you said, we can remove it as a separate ticket.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -0,0 +1,356 @@
+
+	public BufferManager(
+		MemorySegmentProvider globalPool,
+		InputChannel inputChannel,
```
yeah, I also noticed this cycle dependency issue while implementation. And I also tried to cut it if easy to go. But actually I found it needs pay more efforts, because there are many interactions among `InputChannel` and `BufferManager`.

1. `BufferManager` relies on `BufferPool`, then while `SingleInputGate#setup`, it needs to register the respective `BufferPool` for every `InputChannel#BufferManager`. This is the main concern to bypass this issue at current implementation.

2. We also need to maintain the  separate `isReleased` variable inside `BufferManager` to not rely on `InputChannel#isReleased`.

3. We might need another separate interface which would be implemented by `InputChannel`, then we can decouple another two depending methods `InputChannel#onError` and `InputChannel#notifyBufferAvailable`.

pnowojski replied:
I played a bit with this code and maybe it's not perfect, but it's at least a bit better (after separating `BufferManager` from `RemoteInputChannel`), so we can keep it as it is. I don't know how to fix it properly though.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -0,0 +1,356 @@
+			numRequiredBuffers = numRequired;
+
+			while (bufferQueue.getAvailableBufferSize() < numRequiredBuffers && !isWaitingForFloatingBuffers) {
+				BufferPool bufferPool = inputChannel.inputGate.getBufferPool();
```
I also hope so, but the truth is not, because we can not get the proper `BufferPool` while constructing the `BufferManager`. 

The `BufferPool` is created in `SingleInputGate#setup`, and it is also the main reason of relying on `InputChannel` component in `BufferManager` to bypass this issue.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/BufferManager.java:
```diff
@@ -65,14 +65,10 @@
+	public BufferManager(InputChannel inputChannel, int numRequiredBuffers) {
 		checkArgument(numRequiredBuffers >= 0);
+		this.inputChannel = checkNotNull(inputChannel);
+		this.globalPool = inputChannel.inputGate.getMemorySegmentProvider();
```
Actually I injected it in the constructor in previous version, but since we can not get ride of `InputChannel` completed in `BufferManager` as above said, then I reduce this unnecessary argument in constructor.

pnowojski replied:
I think it would be better to limit the reliance on this faulty cyclic dependency. That way it will be easier to get rid of it in the future.

zhijiangW replied:
Makes sense

zhijiangW commented on flink-tests/src/test/java/org/apache/flink/test/streaming/runtime/BackPressureITCase.java:
```diff
@@ -94,7 +94,7 @@ private static Configuration createNetworkBufferConfiguration() {
 
 		final int memorySegmentSizeKb = 32;
-		final MemorySize networkBuffersMemory = MemorySize.parse(memorySegmentSizeKb * (NUM_TASKS + 2) + "kb");
+		final MemorySize networkBuffersMemory = MemorySize.parse(memorySegmentSizeKb * 6 + "kb");
```
In previous way the `LocalBufferPool` for `SingleInputGate` has 0 required buffers, but now we adjust it to guarantee at-least one required buffer for local channel state recovery.

In this ITCase,  the exclusive buffers for map and sink vertex should be 2 * 2 , and the floating buffers in `LocalBufferPool` should be 2 * 1, then the total minimum buffer amount should be 6.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -87,10 +90,14 @@ public LocalInputChannel(
+		// In most cases we only need one buffer for reading recovered state except for very large record.
+		// Then only one floating buffer is required. Even though we need more buffers for recovery for
+		// large record, it only increases some interactions with pool.
+		this.bufferManager = new BufferManager(this, 1);
```
Yes, I agree it is a bit hard to understand. 

The `numRequiredBuffers` factor is only the complex point of this buffer manager model for getting a bit optimization. Actually we can give any initial value for `numRequiredBuffers` (e.g. 0) for unifying the local and remote channels.  And ideally we should adjust this value based on how many total channel states are under unspilling exactly like the concept of `backlog` in credit-based mode.

Actually any value for `numRequiredBuffers` can work correctly now and the only cost is increasing some unnecessary interactions between `BufferManager` and `LocalBufferPool`.
I am really a bit torn here when implementation whether to retain this optimization.

pnowojski replied:
What do you mean by:
> Actually any value for numRequiredBuffers can work correctly now and the only cost is increasing some unnecessary interactions between BufferManager and LocalBufferPool.

? That with `numRequiredBuffers = 0`, the only difference would be that there are not floating buffers initially assigned and they need to be ramped up as the backlog increases?

zhijiangW replied:
I mean we can give any initial `numRequiredBuffers` actually, e.g. `0`. Then while requesting floating buffers from buffer pool in practice, it might have two different paths:

- Get the available buffer right now from buffer pool.

- No available buffer from buffer pool now, then it will register listener to buffer pool for later notification. For this process we should give a proper value of `numRequiredBuffers`. As long as this value is more than 0, it can work correctly.

It is better to give some detail examples to explain how `numRequiredBuffers` works differently in practice. 

E.g. `numRequiredBuffers` is set to `1` for `RemoteInputChannel`. When the exclusive buffers are processed and recycled back to `RemoteInputChannel`, then the current available mount might be 2 which is more than the `numRequiredBuffers (1)`. So when the floating buffers are processed and recycled to `LocalBufferPool`, then it will notify the listener (`RemoteInputChannel`) of available buffers. But the `RemoteInputChannel` already has enough buffers now, so it will not accept this floating buffer to cause waste efforts of register & notify.

But if we give a larger `numRequiredBuffers`,  then we only need to register listener once. After that as long as the floating buffers are processed and recycled, they will be notified to `RemoteInputChannel` and be accepted to reuse future. So it might reduce more interactions between `RemoteInputChannel` and `LocalBufferPool` in practice. Especially for the channel state recovery case, it makes sense to try to occupy all the floating buffers from `LocalBufferPool` because only one channel is unspilling at the same time.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -168,6 +175,12 @@ public void run() {
 	Optional<BufferAndAvailability> getNextBuffer() throws IOException, InterruptedException {
 		checkError();
 
+		BufferAndAvailability bufferAndAvailability = getNextRecoveredStateBuffer();
```
I am neutral for this option, because the similar way really existed in many other places before. E.g. we have `BufferStorage` in `CheckpointedInputGate` for caching the blocked channels' buffers, then while `getNextBuffer` we also need to check whether there are  any pending buffers to be read from `BufferStorage` firstly.

I absolutely agree that it would be better to not have different paths, but I also think it is not so bad if no other easy options.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -168,6 +175,12 @@ public void run() {
 		checkError();
 
+		BufferAndAvailability bufferAndAvailability = getNextRecoveredStateBuffer();
+		if (bufferAndAvailability != null) {
```
My previous assumption was that the local channel will not be chosen by `SingleInputGate` to read if there were no buffers to insert into `RecoveredInputChannel#receivedBuffers` to notify `SingleInputGate.notifyChannelNonEmpty` before.
Or I missed some other corner case?

pnowojski replied:
Generally speaking there are no guarantees that data notifications will be accurate or not (I'm not sure if that's the case here). But the code looks a bit strange here and it relays on some extra assumptions.

zhijiangW replied:
I know that the current data notification from downstream side is not very accurate which might cause return `null` buffer sometimes, but we can make the channel state notification accurate (actually it is the ground truth now). Then another truth is that the channel state consumption should always happen before downstream consumption, so it should be no problem here. 

As long as `getNextBuffer` is triggered by channel state buffer, it can always get accurate buffer. Otherwise if it is triggered by downstream side notification, then it also confirms that the channel state recovery should already finish before.

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannel.java:
```diff
@@ -371,7 +373,14 @@ boolean isWaitingForFloatingBuffers() {
-		return receivedBuffers.poll();
+		synchronized (receivedBuffers) {
+			return receivedBuffers.poll();
+		}
```
it is out of date and i can not find this code path now?

pnowojski replied:
I think it's in a separate commit now `org.apache.flink.runtime.io.network.partition.consumer.RemoteInputChannel#getNumberOfQueuedBuffers`

zhijiangW commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/LocalInputChannel.java:
```diff
@@ -97,6 +100,18 @@ public LocalInputChannel(
+			numRequiredBuffers = 1;
+		}
+
+		super.readRecoveredState(reader);
```
what is the issue here?

pnowojski replied:
Ops, sorry, that was some older comment that I forgot to remove before publishing the review :)

zhijiangW commented:
Thanks for the further reviews @pnowojski ! 

> I'm still not sure if this is the right approach. It's better compared to the original proposal, at a cost of much more added lines of code, while the fundamental issues still remain the same: input channels are even more complicated compared to master branch. Now input channels are interconnected with BufferManager (both Local/RemoteInputChannel and BufferManager are calling/accessing one another many times during a single method invocation) and they are intertwined with RecoveredInputChannel.

Actually I am also a bit torn while implementation and I am also not quite sure which option is the best approach, unless we can fully implement every option to compare through. When I tried to implement the option of working on `CheckpointedInputGate`, I also encountered with some troubles/confusing in other different aspects to make me quit finally.

pnowojski commented on flink-runtime/src/test/java/org/apache/flink/runtime/io/network/partition/consumer/RemoteInputChannelTest.java:
```diff
@@ -731,33 +726,29 @@ public void testFailureInNotifyBufferAvailable() throws Exception {
+			// the unannounced credit if the channel has not requested partition
 			buffer.recycleBuffer();
-			buffer = null;
-			try {
-				failingRemoteIC.checkError();
-				fail("The input channel should have an error based on the failure in RemoteInputChannel#notifyBufferAvailable()");
-			} catch (IOException e) {
-				assertThat(e, hasProperty("cause", isA(IllegalStateException.class)));
-			}
```
This seems like we are loosing a test coverage here? However this test was quite fragile in the first place and I'm not entirely sure what is it suppose to test.

zhijiangW replied:
The previous purpose was to verify that if one remote channel has not requested partition, then the notify credit available will fail with `IllegalStateException` because of non-initialized `PartitionRequestClient` in `RemoteInputChannel`.  I think this test is not meaningful in practice, because if the partition request have not happened yet, then the `onSenderBacklog` should also not happen as well. It seems not valid to explicitly call `onSenderBacklog` to verify this logic without partition request.

My changes for this test is to verify that if the partition request have not happened yet, then the announced credit would not be increased as a result because this is only for the state recovery process to notify available buffers, but this available buffer is not regarded as credit concept without partition request. So I think it makes more sense than the previous test purpose.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -464,6 +472,20 @@ protected void beforeInvoke() throws Exception {
+					inputGate.readRecoveredState(channelIOExecutor, getEnvironment().getTaskStateManager().getChannelStateReader());
+				}
+
+				// Note that we must request partition after all the single gate finishes recovery.
+				for (InputGate inputGate : inputGates) {
+					inputGate.requestPartitions(channelIOExecutor);
+				}
```
It would simplify a threading model, if this was executed from the main thread, via mailbox (in your PR there is already a race condition between processing data (receiving `EndOfPartitionEvent`/releasing resources from the main thread, vs `channelIOExecutor` requesting partitions). 

We could enqueue in the `channelIOExecutor` a simple job, that would enqueue a mail into the mailbox OR `inputGate.readRecoveredState` could return `Future`, and we could enqueue a mail into the mailbox once all futures are completed.

zhijiangW replied:
Yeah, actually I also considered the way of requesting partition by mailbox thread after all futures completed returned by `inputGate.readRecoveredState`.

But I also thought of another existing case to execute partition request by non-task main thread. During `SingleInputGate#updateInputChannel`, when the unknown channel transform into local or remote channel, then it would request partition directly by rpc thread.  So this case makes sense, then my assumption was that partition request actually can be executed by any other threads without race condition issues. So I take the current way instead to save some efforts.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java:
```diff
@@ -46,15 +49,19 @@
 			result = internalReaderRecoveredState(reader, buffer);
 		}
-		getBufferManager().releaseFloatingBuffers();
+		bufferManager.releaseFloatingBuffers();
```
Maybe add
```
LOG.debug("{}/{} Finished recovering input.", inputGate.getOwningTaskName(), channelInfo);
```
?

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -477,14 +477,18 @@ protected void beforeInvoke() throws Exception {
-					inputGate.requestPartitions(channelIOExecutor);
-				}
+				// Note that we must request partition after all the single gates finished recovery.
+				CompletableFuture.allOf(futures).thenRun(ThrowingRunnable.unchecked(() -> {
+					for (InputGate inputGate : inputGates) {
+						inputGate.requestPartitions();
+					}
+				}));
```
Hmmm, this is a bit fragile, as it implicitly assumes futures are completed from the task thread? Maybe add a `checkState(...)` asserting a thread to document this assumption?

zhijiangW replied:
Yeah I assume it should be completed by task thread, actually my initial version was `mailboxProcessor.getMainMailboxExecutor().execute()`.

I can add a `checkState` here to guarantee that the current thread is mailbox thread and then give some note comments.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/RecoveredInputChannel.java:
```diff
@@ -177,17 +178,18 @@ void sendTaskEvent(TaskEvent event) {
-			releasedBuffers.addAll(receivedBuffers);
-			receivedBuffers.clear();
-			isReleased = true;
+		if (isReleased.compareAndSet(false, true)) {
```
Why atomic? It could easily be checked and set under `synchronized (receivedBuffers)` lock, simplifying threading model a bit and it would also avoid extra `AtomicBoolean` check on the hot path.

zhijiangW replied:
In `RemoteInputChannel` we also had the atomic released variable and I guess here is the similar case. The release operation might be called not only by task thread, but also by canceler thread. If the latter, it might has visibility issue for task thread while interacting with `RecoveredInputChannel#isReleased` in `BufferManager` related operations.

pnowojski replied:
> also by canceler thread. If the latter, it might has visibility issue 

it wouldn't:

>  It could easily be checked and set under `synchronized (receivedBuffers)` lock

?

zhijiangW replied:
yeah, we can also take the way of adding the `synchronized (receivedBuffers)` for the  method of `RecoveredInputGate#isReleased()`

zhijiangW replied:
It seems we have different ways for this issue in different classes. `LocalInputChannel` introduces `volatile` for this variable, and `RemoteInputChannel` introduces `atomic` variable, but `RecoveredInputChannel` takes the normal boolean variable. 

We might need to unify them in separate commits future.

pnowojski replied:
Agreed. But generally speaking closing/releasing logic is on my list for a larger refactor/clean up for a long time, especially because of those various nasty concurrency issues - I would see closing/releasing eventually moved to task thread/mailbox.

zhijiangW commented:
Thanks for quick review @pnowojski ! 

Considering the issue of `requestPartition` in `AbstractRecordReader#getNextRecord`
> Not sensitive for batch 

I mean the batch job might not be such performance sensitive compared with streaming mode, furthermore we also had this action long time ago without mailbox model if I remembered correctly. I can double check whether this action can be done only once in different code paths.

zhijiangW commented:
I have addressed the above comments in new separate commits and squash all the previous commits. @pnowojski

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/api/reader/AbstractRecordReader.java:
```diff
@@ -63,6 +63,8 @@ protected AbstractRecordReader(InputGate inputGate, String[] tmpDirectories) {
 	}
 
 	protected boolean getNextRecord(T target) throws IOException, InterruptedException {
+		inputGate.requestPartitions();
```
nit: maybe add a small comment why is done like so here?

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/consumer/SingleInputGateFactory.java:
```diff
@@ -190,6 +190,7 @@ private InputChannel createInputChannel(
 					partitionRequestMaxBackoff,
 					metrics);
 			},
+			// TODO: can we sometimes call createKnownInputChannel?
```
nit: remove comment

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
@flinkbot run azure

zhijiangW commented:
@pnowojski I already addressed the last nit comments. Besides that I also adjusted another two places. 

1. In `StreamTask#beforeInvoke`, I take the following way instead to guarantee the partition request execute by task thread.

```
CompletableFuture.allOf(futures).thenRun(() -> mailboxProcessor.getMainMailboxExecutor().execute(
					() -> {
						for (InputGate inputGate : inputGates) {
							inputGate.requestPartitions();
						}
					}, "Input gates request partitions"));
```

2. In `RecoveredInputChannel#getNextRecoveredStateBuffer`, I move the future complete and related processes out of the `synchronized (receivedBuffers)` to avoid potential deadlock issue.

I will merge the PR after azure pass if you have no further concerns.

pnowojski commented:
LGTM. One question:

> In RecoveredInputChannel#getNextRecoveredStateBuffer, I move the future complete and related processes out of the synchronized (receivedBuffers) to avoid potential deadlock issue.

Was this a precaution or a fix to a real deadlock? I don't see how in this case completing the future from within the critical section could deadlock, as the only thing attached it enqueuing something in the mailbox?

zhijiangW commented:
In previous way it already had the deadlock case verified by `FastFailuresITCase`. 

1. The future complete is done in task thread while `RecoveredInputChannel#getNextRecoveredStateBuffer`, then it is within the synchronized lock `receivedBuffers`. Then the task thread would execute partition request immediately. During partition request, it would occupy `SingleInputGate#requestLock`.

2. At the meanwhile, the canceler thread might call `SingleInputGate#close` to occupy `requestLock` firstly. Then when it calls `RecoveredInputChannel#releaseAllResources`, it needs to synchronized  the lock of `RemoteInputChannel#receivedBuffers`. Note that this lock was already occupied by task thread in above first step.

3. So the canceler thread would wait for the task thread to free `RemoteInputChannel#receivedBuffers` lock, and the task thread is waiting for the canceler thread free `SingleInputGate#requestLock`.

Actually it only needs one of my above two changes to resolve this deadlock, either completing the future out of `RemoteInputChannel#receivedBuffers` lock, or add the partition request mail in `StreamTask` instead to avoid trying to lock `SingleInputGate#requestLock` immediately. 

Finally I made both changes because:
1. complete the future actually does not need the lock, so better to do it outside.
2. Enqueue the mail into mailbox can avoid checking whether the future is completed by task thread or not.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamTask.java:
```diff
@@ -458,17 +466,50 @@ protected void beforeInvoke() throws Exception {
 
+	private void readRecoveredChannelState() throws IOException, InterruptedException {
+		//TODO we will support channel state recovery even if the current setting is not unaligned checkpoint.
+		if (!configuration.isUnalignedCheckpointsEnabled()) {
```
It will be addressed in FLINK-17476
